logs: remove deprecated klog flags

pohly opened the pull request:
#### What type of PR is this?
/kind cleanup
/kind api-change

#### What this PR does / why we need it:

This completes the deprecation of klog flags which are no longer supported.
klog itself continues to support them, but Kubernetes components don't. This
makes the command line interfaces simpler and reduces the attack surface
because less functionality is exposed.

#### Which issue(s) this PR fixes:
Fixes #https://github.com/kubernetes/enhancements/issues/2845

#### Special notes for your reviewer:

For example, kube-controller-manager now has:

    Logs flags:

      --log-flush-frequency duration
                Maximum number of seconds between log flushes (default 5s)
      --log-json-info-buffer-size quantity
                [Alpha] In JSON format with split output streams, the info messages can be buffered for a while to increase performance. The default value of zero
                bytes disables buffering. The size can be specified as number of bytes (512), multiples of 1000 (1K), multiples of 1024 (2Ki), or powers of those (3M,
                4G, 5Mi, 6Gi). Enable the LoggingAlphaOptions feature gate to use this.
      --log-json-split-stream
                [Alpha] In JSON format, write error messages to stderr and info messages to stdout. The default is to write a single stream to stdout. Enable the
                LoggingAlphaOptions feature gate to use this.
      --logging-format string
                Sets the log format. Permitted formats: "json" (gated by LoggingBetaOptions), "text".
       -v, --v Level
                number for the log level verbosity
      --vmodule pattern=N,...
                comma-separated list of pattern=N settings for file-filtered logging (only works for text log format)

    Misc flags:

      --kubeconfig string
                Path to kubeconfig file with authorization and master location information.
      --master string
                The address of the Kubernetes API server (overrides any value in kubeconfig).

    Global flags:

     -h, --help
                help for kube-controller-manager
      --version version[=true]
                Print version information and quit

For details see
https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components


#### Does this PR introduce a user-facing change?
```release-note
Legacy klog flags are no longer available. Only `-v` and `-vmodule` are still supported.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/2845
- [Usage]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components/README.md#user-stories
```

/assign @serathius

pohly committed:
logs: remove deprecated klog flags

This completes the deprecation of klog flags which are no longer supported.
klog itself continues to support them, but Kubernetes components don't. This
makes the command line interfaces simpler and reduces the attack surface
because less functionality is exposed.

For example, kube-controller-manager now has:

    Logs flags:

      --log-flush-frequency duration
                Maximum number of seconds between log flushes (default 5s)
      --log-json-info-buffer-size quantity
                [Alpha] In JSON format with split output streams, the info messages can be buffered for a while to increase performance. The default value of zero
                bytes disables buffering. The size can be specified as number of bytes (512), multiples of 1000 (1K), multiples of 1024 (2Ki), or powers of those (3M,
                4G, 5Mi, 6Gi). Enable the LoggingAlphaOptions feature gate to use this.
      --log-json-split-stream
                [Alpha] In JSON format, write error messages to stderr and info messages to stdout. The default is to write a single stream to stdout. Enable the
                LoggingAlphaOptions feature gate to use this.
      --logging-format string
                Sets the log format. Permitted formats: "json" (gated by LoggingBetaOptions), "text". (default "text")
  -v, --v Level
                number for the log level verbosity
      --vmodule pattern=N,...
                comma-separated list of pattern=N settings for file-filtered logging (only works for text log format)

    Misc flags:

      --kubeconfig string
                Path to kubeconfig file with authorization and master location information.
      --master string
                The address of the Kubernetes API server (overrides any value in kubeconfig).

    Global flags:

  -h, --help
                help for kube-controller-manager
      --version version[=true]
                Print version information and quit

For details see
https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components

pohly commented:
@serathius: do we still want to have this line here?

```
Non-default formats don't honor these flags: --vmodule.
```

It only lists one flag, and that flag itself now says "only works for text log format".

serathius commented:
/lgtm

pohly commented:
/hold

I forgot to mention that https://github.com/kubernetes/test-infra/pull/27305 needs to be merged first.

pohly commented:
@serathius so do you want me to remove "Non-default formats don't honor these flags: --vmodule." or keep it?

serathius commented:
I don't think `"Non-default formats don't honor these flags: --vmodule."` is needed, we can remove it.

yuzhiquan commented on staging/src/k8s.io/component-base/logs/klogflags/klogflags.go:
```diff
@@ -0,0 +1,42 @@
+/*
+Copyright 202 The Kubernetes Authors.
```
```suggestion
Copyright 2022 The Kubernetes Authors.
```

pohly replied:
Oh, the good old times of 202! You are right of course, good catch.

pohly commented:
> I don't think `"Non-default formats don't honor these flags: --vmodule."` is needed, we can remove it.

That's also my preference. While at it, I also removed "Non-default choices are currently alpha and subject to change without warning." because that has been wrong for a while now.

We don't need that sentence because non-GA formats list the corresponding feature gate:
```
          --logging-format string
                    Sets the log format. Permitted formats: "json" (gated by LoggingBetaOptions), "text". (default "text")
```

pohly committed:
stop using deprecated klog flags

Some scripts and tools still relied on the deprecated flags, the ones
which are about to be removed.

This is intentionally not a complete removal of all those flags in the entire
repo. This would lead to much more code churn also in places where commands
still accept the flags because they use klog directly.

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
@serathius: can you take another look? All tests should be passing now after removing some more occurrences of the deprecated flags.

serathius commented:
/lgtm

pohly commented:
/assign @klueska

Can you approve for SIG Node?

pohly commented on test/e2e_node/jenkins/conformance/conformance-jenkins.sh:
```diff
@@ -33,7 +33,7 @@ TIMEOUT=${TIMEOUT:-"45m"}
 
 go run test/e2e_node/runner/remote/run_remote.go  --test-suite=conformance \
-  --logtostderr --vmodule=*=4 --ssh-env="gce" --ssh-user="$GCE_USER" \
+  --vmodule=*=4 --ssh-env="gce" --ssh-user="$GCE_USER" \
```
None. Logging to stderr has been the default for several years now, so using this parameter didn't change anything.

klueska replied:
Sounds good. Thanks for clarifying.

klueska commented:
/approve
For node_e2e changes

pohly commented:
/assign @thockin 

For approval.

dashpole commented:
/triage accepted

SergeyKanzhelev commented:
/assign @rphillips

SergeyKanzhelev commented:
/priority important-longterm

thockin commented on staging/src/k8s.io/component-base/logs/logs.go:
```diff
@@ -47,7 +42,7 @@ var (
 
 func init() {
-	klog.InitFlags(packageFlags)
+	klogflags.Init(packageFlags)
```
I'm unclear why this moved to a different package, personally.

pohly replied:
It's used in `k8s.io/component-base/logs/api/v1` to access the klog settings and determine which logging flags exist, which in turn is used to warn about those that only have an effect in the text format (used to more, now it's only `-vmodule` which isn't supported by zapr).

If it was not moved into a separate leaf package, there would be a circular dependency. Before, `api/v1` directly called `klog.InitFlags`, but that's not correct anymore because only a subset of those flags are active.

pohly replied:
Perhaps I also need to explain why klogflags is needed and why it is under component-base:
- We cannot simply remove the flags from klog.InitFlags, that would break non-Kubernetes applications that use it.
- We could add a klog.InitKubernetesFlags, but IMHO that is too specific to Kubernetes and belongs more into component-base than klog.

thockin reviewed (commented):
I think this all looks sane

thockin commented:
/approve
